make HiPE work on x86_64 when PIE is enabled

bhuztez opened the pull request:
Currently HiPE amd64 assumes the runtime system code is loaded into
the low 2G of the address space. However, this is not the case when
PIE is enabled, it is loaded into a random location. So trampolines
are required to call BIFs, and also we have first to load the address
of sse2_fnegate_mask to a regisiter before xorpd in fchs.

-----------------------

@margnus1 @mikpe @sverker

Since #1655 is closed, I create a new pull request on this. I tried the trampoline approach a few months ago, however, it does not work because `sse2_fnegate_mask` is not in the low 2G of the address space. Then I found `erts/emulator/hipe/hipe_amd64_abi.txt`, which said the small code model is chosen to avoid trampolines.

Thus, I tried a different approach, which put a dispatch table and `sse2_fnegate_mask` at the end of each module to avoid changes to `lib/hipe/x86/hipe_rtl_to_x86.erl`. This is a little bit awkward, since the address returned by `erts_alloc` is 8-bytes aligned on x86-64, and `sse2_fnegate_mask` has to be 16-bytes aligned. Moreover, only very recent versions of LLVM, [Support nonlazybind attribute for X86 64-bit ELF](https://reviews.llvm.org/D39065).

So right now, trampolines and changes to `lib/hipe/x86/hipe_rtl_to_x86.erl` are inevitable without big changes to other parts of the code.

sverker commented on erts/emulator/hipe/hipe_amd64.c:
```diff
@@ -96,12 +102,70 @@ static void *alloc_code(unsigned int alloc_bytes)
+    Uint i;
+
+    if (is_not_tuple(callees))
+	return -1;
```
check_callees() must accept NIL meaning no callees and return 0.

bhuztez replied:
@sverker I copied this from `hipe_arm.c` which does not check if it is NIL either.

We got a NIL here, because `lib/kernel/src/hipe_unified_loader.erl` has been changed but bootstrap beam files are left not updated.

sverker commented:
After fixing the callees==NIL problem,
I tested to build with `./otp_build setup --enable-native-libs`
and got the following strange symptom:

```
=== Entering application syntax_tools
make[3]: Entering directory '/home/uabseri/src/otp2/lib/syntax_tools/src'
 ERLC	../ebin/erl_syntax.beam
:
 ERLC	merl_transform.beam
 ERLC	../ebin/merl_transform.beam
merl_transform.erl: undefined parse transform 'merl_transform'
Makefile:74: recipe for target '../ebin/merl_transform.beam' failed
make[3]: *** [../ebin/merl_transform.beam] Error 1
```

bhuztez commented:
@sverker However, after running `./otp_build update_primary`, there are some changes left uncommitted.

```
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   bootstrap/lib/compiler/ebin/compiler.app
	modified:   bootstrap/lib/compiler/ebin/compiler.appup
	modified:   bootstrap/lib/kernel/ebin/kernel.app
	modified:   bootstrap/lib/kernel/ebin/kernel.appup
	modified:   bootstrap/lib/stdlib/ebin/stdlib.app
	modified:   bootstrap/lib/stdlib/ebin/stdlib.appup

no changes added to commit (use "git add" and/or "git commit -a")
```

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/erlang/otp?pullRequest=1669) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/erlang/otp?pullRequest=1669) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: sverker<br/>:x: bhuztez<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/erlang/otp?pullRequest=1669) it.</sub>

sverker commented:
You were right, updating primary bootstrap for `hipe_unified_loader` was enought to make `--enable-native-libs` work. The `*.app` and `*.appup` files are irrelevant for bootstrap execution.

I pushed a commit with hipe_unified_loader.beam to your hipe-amd64-trampoline.

sverker commented on erts/emulator/hipe/hipe_amd64.c:
```diff
@@ -71,14 +72,19 @@ int hipe_patch_insn(void *address, Uint64 value, Eterm type)
 
-    ASSERT(trampoline == NULL);
+    if ((destOffset < -0x80000000L) || (destOffset >= 0x80000000L)) {
+        destOffset = (Sint64)trampoline - (Sint64)callAddress - 4;
```
This does not work. The trampoline code starts at `(trampoline-2)`.

sverker commented:
@bhuztez Does this mean you have actually tested it now on a PIE configuration?

bhuztez commented:
@sverker Yes. I guess last time, I forgot to run configure again.

This time

```console
$ git clean -d -f -X
$ export ERL_TOP="$(pwd)"
$ ./otp_build autoconf
$ ./configure --enable-native-libs CFLAGS='-fPIE' LDFLAGS='-pie'
$ make
$ cd release/tests/test_server
$ "${ERL_TOP}/bin/erl"
1> ts:install().
2> ts:run(emulator, hipe_SUITE, [batch]).
3> ts:run(hipe, [batch]).
```
After that, I run `readelf -h bin/x86_64-unknown-linux-gnu/beam.smp`
```
ELF Header:
  Magic:   7f 45 4c 46 02 01 01 00 00 00 00 00 00 00 00 00 
  Class:                             ELF64
  Data:                              2's complement, little endian
  Version:                           1 (current)
  OS/ABI:                            UNIX - System V
  ABI Version:                       0
  Type:                              DYN (Shared object file)
  Machine:                           Advanced Micro Devices X86-64
  Version:                           0x1
  Entry point address:               0x65ac0
  Start of program headers:          64 (bytes into file)
  Start of section headers:          4607952 (bytes into file)
  Flags:                             0x0
  Size of this header:               64 (bytes)
  Size of program headers:           56 (bytes)
  Number of program headers:         9
  Size of section headers:           64 (bytes)
  Number of section headers:         32
  Section header string table index: 31
```

Then I removed the line `flags |= MAP_32BIT` in `erts/emulator/sys/common/erl_mmap.c`, and tried again.

sverker commented:
Rebased, squashed and merged to master at 1fe1c9a57d365c8203cb1380e4bdfe8f7f6ab8ce.
